### PR TITLE
Admin panel

### PR DESCRIPTION
**Pull request by tyxchen**
Add support for the front-end admin panel. Includes new methods/routes for managing users.

### CODE REVIEW

File: user/models.py [hunk 1]
```diff
@@ -1,23 +1,111 @@
...
+        if 'email' in data:
+            try:
+                validate_email(data['email'])
+            except ValidationError as err:
+                errors_list['email'] = ['USER.EMAIL.INVALID']
+
+        if 'password' in data:
+            try:
+                user = self.instance or SluglineUser(**data)
+                password_validators = (
```
Comment (ktrieu) [inline user/models.py]: Believe we can set these in the settings system wide.

File: user/views.py [hunk 2]
```diff
@@ -37,4 +53,113 @@ def auth_view(request):
             'authenticated': True,
             'user': UserSerializer(request.user).data
         })
-    
+
+
+@api_view(['GET'])
```
Comment (ktrieu) [inline user/views.py]: We can replace all these with a ViewSet.

File: content/management/commands/create_groups.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+        editor.permissions.add(Permission.objects.get(name='Can change article'))
+        editor.permissions.add(Permission.objects.get(name='Can delete article'))
+        editor.permissions.add(Permission.objects.get(name='Can view article'))
+        editor.permissions.add(Permission.objects.get(name='Can add issue'))
+        editor.permissions.add(Permission.objects.get(name='Can change issue'))
+        editor.permissions.add(Permission.objects.get(name='Can delete issue'))
+        editor.permissions.add(Permission.objects.get(name='Can view issue'))
+
+        contrib.permissions.add(Permission.objects.get(name='Can add article'))
+        contrib.permissions.add(Permission.objects.get(name='Can change article'))
```
Comment (ktrieu) [inline content/management/commands/create_groups.py:29]: Less a review than a question: how do we plan on restricting people to only editing their own articles? Can the permissions system handle this or are we going to have to do this ourselves?

File: user/urls.py [hunk 4]
```diff
@@ -1,9 +1,16 @@
...
+from user.views import *
 
 urlpatterns = [
     path('login/', login_view),
     path('logout/', logout_view),
-    path('auth/', auth_view)
+    path('auth/', auth_view),
+    path('user/update', update_user_view),
+    path('allusers', list_users_view),
+    path('createuser', create_user_view),
```
Comment (ktrieu) [inline user/urls.py]: We can replace a lot of these with a router tied to that ViewSet I mentioned below this.

File: user/models.py (hunk 1 above)
Comment (tyxchen) [reply user/models.py]: 1d2b3023

File: user/views.py (hunk 2 above)
Comment (tyxchen) [reply user/views.py]: 9b4bbca2

File: content/management/commands/create_groups.py (hunk 3 above)
Comment (tyxchen) [reply content/management/commands/create_groups.py:29]: That's a good question...and I'm not too sure how to answer that. But in the future one feature I'd like to implement is sharing and collaborative editing of articles, so it's probably best to handle edit/view permissions ourselves.

File: user/urls.py (hunk 4 above)
Comment (tyxchen) [reply user/urls.py]: 9b4bbca2

**Comment by tyxchen** [comment]
I have a few complementary changes to 9b4bbca2 and d281482b in slugline-web that I've started a PR for (uwmathnews/slugline-web#2).

File: content/management/commands/create_groups.py (hunk 3 above)
Comment (ktrieu) [reply content/management/commands/create_groups.py:29]: Alright, we'll get there when we get there.

File: user/models.py [hunk 5]
```diff
@@ -1,23 +1,107 @@
...
 
+FORBIDDEN_USERNAMES = {'admin', 'administrator', 'root', 'toor', 'sudo', 'sudoers'}
+
+
+class SluglineUser(AbstractUser):
+    email = models.EmailField(blank=False)
     """Articles written by this user will use this name by default."""
     writer_name = models.CharField(max_length=255)
 
+    @property
```
Comment (ktrieu) [inline user/models.py:19]: Is there any particular reason this has to be a property considering its read only?

File: user/models.py [hunk 6]
```diff
@@ -1,23 +1,107 @@
...
+            except ValidationError as err:
+                errors_list['email'] = ['USER.EMAIL.INVALID']
+
+        if 'password' in data:
+            try:
+                current_attrs = UserSerializer(self.instance).data
+                current_attrs.update(data)
+                validate_password(data['password'], user=SluglineUser(**current_attrs))
+            except ValidationError as err:
+                print(list(map(lambda e: (e.code, e.params), err.error_list)))
```
Comment (ktrieu) [inline user/models.py]: Leftover print statement

File: user/views.py [hunk 7]
```diff
@@ -1,40 +1,158 @@
...
-from user.models import UserSerializer
+from user.models import SluglineUser, UserSerializer, FORBIDDEN_USERNAMES
+
+
+class IsEditor(BasePermission):
+    def has_permission(self, request, view):
+        return bool(request.user and (request.user.is_editor or request.user.is_staff))
+
+
+class SluglineAPIException(APIException):
```
Comment (ktrieu) [inline user/views.py]: Should move this to a common file somewhere.

File: user/views.py [hunk 8]
```diff
@@ -1,40 +1,158 @@
...
+from rest_framework.decorators import api_view, permission_classes, action
+from rest_framework.permissions import BasePermission, IsAuthenticated
 from rest_framework.response import Response
 from rest_framework.exceptions import APIException
 
-from user.models import UserSerializer
+from user.models import SluglineUser, UserSerializer, FORBIDDEN_USERNAMES
+
+
+class IsEditor(BasePermission):
```
Comment (ktrieu) [inline user/views.py]: Should move this to a common file somewhere.

File: user/urls.py [hunk 9]
```diff
@@ -1,9 +1,15 @@
...
+from user.views import *
+
+router = SimpleRouter()
+router.register('users', UserViewSet, 'users')
 
 urlpatterns = [
     path('login/', login_view),
     path('logout/', logout_view),
-    path('auth/', auth_view)
+    path('user/', retrieve_user_view),
```
Comment (ktrieu) [inline user/urls.py]: Think having `user/` and `users/` as two subtly different URLs is a bad idea, maybe call the current user one `me/`.

File: user/models.py (hunk 5 above)
Comment (tyxchen) [reply user/models.py:19]: I wanted conformity with is_authenticated, hence the property. You're right on the setter though, I'll remove that.

Comment (tyxchen) [reply user/models.py:19]: cdfaccba

File: user/models.py (hunk 6 above)
Comment (tyxchen) [reply user/models.py]: cdfaccba

File: user/views.py (hunk 7 above)
Comment (tyxchen) [reply user/views.py]: 26c6edaa

File: user/views.py (hunk 8 above)
Comment (tyxchen) [reply user/views.py]: 26c6edaa

File: user/urls.py (hunk 9 above)
Comment (tyxchen) [reply user/urls.py]: e2e58a29

**Review by ktrieu — APPROVED** [review]

**Pull request merged**